### PR TITLE
[feat] Add Keycloak Scope Client CRUD

### PR DESCRIPTION
**Pull request by leebongseung**
## 📝 Work Done
- [x] Added `KeycloakAuthScopeAsyncClient` interface (CRUD methods)
- [x] Implemented `KeycloakAuthScopeAsyncClientImpl` with WebFlux
- [x] Added `KeycloakAuthScopeClient` and sync implementation using `block()`
- [x] Updated `ClientConfiguration` with scope URL methods
- [x] Created `ScopeQueryParams` DTO and query string builder
- [x] Updated `KeycloakClient` factory to include scope clients
- [x] Added unit tests (`ScopeQueryParamsTest`, `ClientConfigurationTest`)
- [x] Updated `realm.json` with new scope resources

---

## 💡 Review Points
1. Check if scope CRUD URLs and params in `ClientConfiguration` are correct.
2. Make sure `ScopeQueryParams.toQueryString()` creates clear and correct query strings.
3. Review if using `block()` in sync client is safe and reasonable.
4. See if test cases cover enough edge cases (empty params, invalid values, etc.).

---

## 📂 Changed Files
1. `KeycloakAuthScopeAsyncClient.java`: async client interface
2. `KeycloakAuthScopeAsyncClientImpl.java`: async implementation (CRUD APIs)
3. `KeycloakAuthScopeClient.java`: sync client interface
4. `KeycloakAuthScopeClientImpl.java`: sync implementation (`block()`)
5. `ClientConfiguration.java`: added scope URL methods
6. `ScopeQueryParams.java`: DTO for search params + query builder
7. `KeycloakClient.java`: added scope clients in factory
8. `ClientConfigurationTest.java`: test for scope URLs
9. `ScopeQueryParamsTest.java`: test for query string
10. `realm.json`: added scope resource changes

---

## 🚨 Other Notes
- Need to decide how to handle errors when creating or deleting scopes.
- Mixed use of `Mono` and `block()` may cause issues depending on app runtime (reactive vs blocking).

### CODE REVIEW

File: src/main/java/com/sd/KeycloakClient/client/admin/auth/async/KeycloakAuthScopeAsyncClient.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.sd.KeycloakClient.client.admin.auth.async;
+
+import com.sd.KeycloakClient.dto.KeycloakResponse;
+import com.sd.KeycloakClient.dto.admin.ScopeQueryParams;
+import org.keycloak.representations.idm.authorization.ScopeRepresentation;
+import reactor.core.publisher.Mono;
+
+public interface KeycloakAuthScopeAsyncClient {
```
Comment (sangwonyu) [inline src/main/java/com/sd/KeycloakClient/client/admin/auth/async/KeycloakAuthScopeAsyncClient.java:8]: 인터페이스에 대한 설명 써주면 좋을거같아요! ScopeRepresentation 같은 컬럼에 대한 설명도 있으면 좋을거같네요

File: src/main/java/com/sd/KeycloakClient/client/admin/auth/sync/KeycloakAuthScopeClient.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.sd.KeycloakClient.client.admin.auth.sync;
+
+import com.sd.KeycloakClient.dto.KeycloakResponse;
+import com.sd.KeycloakClient.dto.admin.ScopeQueryParams;
+import org.keycloak.representations.idm.authorization.ScopeRepresentation;
+
+public interface KeycloakAuthScopeClient {
```
Comment (sangwonyu) [inline src/main/java/com/sd/KeycloakClient/client/admin/auth/sync/KeycloakAuthScopeClient.java:7]: 여기도 설명 간단히 적으면 좋을거같아요

File: src/main/java/com/sd/KeycloakClient/client/admin/auth/async/impl/KeycloakAuthScopeAsyncClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+   @Override
+   public Mono<KeycloakResponse<ScopeRepresentation[]>> getScopes(
+       String accessToken,
+       String clientUuid,
+       ScopeQueryParams scopeQueryParams
+   ) {
+
+      return http.<ScopeRepresentation[]>get(configuration.getAuthScopeSearchUrl(clientUuid, scopeQueryParams.toQueryString()))
+          .authorizationBearer(accessToken)
+          .entities(scopeQueryParams)
```
Comment (sangwonyu) [inline src/main/java/com/sd/KeycloakClient/client/admin/auth/async/impl/KeycloakAuthScopeAsyncClientImpl.java:43]: image
api 문서 보면 entities 없는거 같은데 맞을까요?

File: src/test/java/com/sd/KeycloakClient/config/ClientConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      final String expectedUrl =
+          RELATIVE_PATH + "/admin/realms/" + clientConfiguration.getRealmName() + "/clients/" + CLIENT_ID + "/authz/resource-server/scope/"
+              + SCOPE_UUID;
+
+      Assertions.assertEquals(authScopeUrl, expectedUrl);
+   }
+
+   @Test
+   @DisplayName("case3. success case: get auth scope search url with query params")
+   void getAuthScopeSearchUrl() {
```
Comment (sangwonyu) [inline src/test/java/com/sd/KeycloakClient/config/ClientConfigurationTest.java:54]: @ParameterizedTest
활용해서 name 뿐만 아니라 모든 검색 조건 테스트해보면 좋을거같아요

File: src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      @DisplayName("case1. no parameters provided, should return empty query string")
+      void toQueryString_noParams() {
+         ScopeQueryParams scopeQueryParams = new ScopeQueryParams();
+         String queryString = scopeQueryParams.toQueryString();
+
+         assertTrue(queryString.isEmpty());
+      }
+
+      @Test
+      @DisplayName("case2. with parameters provided, should return query string")
```
Comment (sangwonyu) [inline src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java:30]: Nest 사용할거면 사용의 가치에 맞게 조금 더 구분 지어서 다양한 테스트 케이스 넣으면 좋을거같아요. 
예를 들어, 성공, 실패, 경계로 나눠서 하면 테스트 구분.

File: src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      void toQueryString_withParams() {
+         ScopeQueryParams scopeQueryParams =
+             ScopeQueryParams.builder()
+                 .first("0")
+                 .max("100")
+                 .build();
+
+         String queryString = scopeQueryParams.toQueryString();
+
+         String[] parts = queryString.substring(1).split("&");
```
Comment (sangwonyu) [inline src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java:40]: 여기서 이렇게 비즈니스 로직을 넣기 보다는 예측 값을 상수로 비교 테스트 하는게 어떨까요.
실제 운영 로직에 없는 로직을 넣으므로써 의도치 않게 테스트 실패를 가져갈 수 있으니 예측 값 그대로 비교 테스트하는게 좋을거같네요.

File: src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      void toQueryString_noParams() {
+         ScopeQueryParams scopeQueryParams = new ScopeQueryParams();
+         String queryString = scopeQueryParams.toQueryString();
+
+         assertTrue(queryString.isEmpty());
+      }
+
+      @Test
+      @DisplayName("case2. with parameters provided, should return query string")
+      void toQueryString_withParams() {
```
Comment (sangwonyu) [inline src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java:31]: 하나의 케이스만 테스트 하는게 아니라 다양한 테스트 케이스 넣으면 좋을거같아요. 

성공 케이스에는 
모두 넣는 경우

경계 테스트에서는 
값이 없는 경우 각각 하나씩 넣는 경우 경우

위에서 언급하긴 했지만, ParameterizedTest이용하면 수월하게 할 수 있을거같아요

File: src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class ScopeQueryParamsTest {
```
Comment (sangwonyu) [inline src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java:13]: 통합 테스트는 안보이는거같아요!

File: src/main/java/com/sd/KeycloakClient/client/admin/auth/async/impl/KeycloakAuthScopeAsyncClientImpl.java (hunk 3 above)
Comment (leebongseung) [reply src/main/java/com/sd/KeycloakClient/client/admin/auth/async/impl/KeycloakAuthScopeAsyncClientImpl.java:43]: 잘못넣은것같습니다 확인완료했습니다!

File: src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java (hunk 8 above)
Comment (leebongseung) [reply src/test/java/com/sd/KeycloakClient/dto/admin/ScopeQueryParamsTest.java:13]: 외부와의 상호작용(DB, keyclaok 연동 등) 이 존재하지 않아서 단위 테스트만 수행했습니다.
Dto를 통해 외부와 통신하는 부분에서 별도로 통합 테스트 진행하였습니다!

File: src/test/java/com/sd/KeycloakClient/client/admin/auth/async/impl/KeycloakAuthScopeAsyncClientImplTest.java [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.keycloak.representations.idm.authorization.ScopeRepresentation;
+import reactor.core.publisher.Mono;
+import reactor.test.StepVerifier;
+
+@MockKeycloakClient
+class KeycloakAuthScopeAsyncClientImplTest extends KeycloakShareTestContainer {
```
Comment (sangwonyu) [inline src/test/java/com/sd/KeycloakClient/client/admin/auth/async/impl/KeycloakAuthScopeAsyncClientImplTest.java:23]: 위에 에서도 언급하긴 했는데 전반적으로 쿼리 조건을 꼼꼼하게 넣어주면 좋을거같아요. 
쿼리 조건에 대해 각 조건이 다 잘 작동하는지.
여러 조건을 넣었을 때 and 조건이면 and 조건에 맞게 출력이 되는지.
없을 경우도 전체를 조회할지 빈 값을 조회할지 정책 정하고 그 정책에 맞게 출력이 되는지.

Comment (sangwonyu) [reply src/test/java/com/sd/KeycloakClient/client/admin/auth/async/impl/KeycloakAuthScopeAsyncClientImplTest.java:23]: 그리고,  Nest로 묶어서 하면 좀 더 가독성 있고 좋을거같아요

**Review by sangwonyu — APPROVED** [review]
